### ISSUE TITLE
How to prevent element from being removed during morph?

### ISSUE THREAD

**Issue by airblade** [hotwired/turbo#1279]
I have a list of the 10 most recent things on a page, ordered new-to-old.

When a new thing is created, the page refreshes and morphs.  The newly created thing is at the top of the list, while the thing which was at the bottom of the list disappears – as expected.

However I would like to prevent the thing at the bottom of the list from being removed during a morph.  I want creating a new thing to add the thing to the top of the list without dropping anything off the end.

First I tried adding `data-turbo-permanent` to each thing.  That does prevent the things from being removed from the DOM but the list gets shuffled.  I want the list's new-to-old order to be preserved.

I then tried the `turbo:before-morph-element` event.  The [docs](https://turbo.hotwired.dev/reference/events#turbo%3Abefore-morph-element) say that `event.target` is the original element and `event.detail.newElement` is the new element.  I assumed that for an element which is about to be removed by Idiomorph, `event.detail.newElement` would be undefined.  In which case I would call `event.preventDefault()` to prevent the removal.  However console logging shows that `event.detail.newElement` is always undefined – is that expected?

How can I determine if morphing is about to remove an element, and then prevent it from happening?

**Comment by airblade** [comment]
I wonder if looking at the parent element (`<ul id="things">`) would work better.

**Comment by airblade** [comment]
I figured this out:

- Exempt each `<li>` from morphing with `data-turbo-permanent`.
- The "append" steam action [removes elements in common](https://github.com/hotwired/turbo/blob/c339144924e5632d4ef5e3a7b941b3db17d9efa3/src/core/streams/stream_actions.js#L10) before appending all the incoming ones.  That is why the order changes.
- Normally to control the order of children in a container, you would re-render the container.  But because the container could have accumulated many items subsequently via cursor-pagination, I don't want to re-render all of them: it would defeat the purpose of pagination.
- So to get the order I want I added a custom stream action to sort them after they have been appended.
